### PR TITLE
Update and not add

### PR DESCRIPTION
**Pull request by matanel-oren**
For issue #8

Changed:
When we want to add a group, if there exists CourseGroup with the same (CourseOccurence, class_type, mark), this group will be updated and wont be added along side.
For the classes: those who didn't changed will not be changed, those who don't exist in the db will be added, and those who exist and not to be added - they are irrelevant and will be deleted.

### CODE REVIEW

File: academic_helper/models/course_occurrence.py [hunk 1]
```diff
@@ -92,10 +92,20 @@ class ClassGroup(Base):
     occurrence = models.ForeignKey(CourseOccurrence, on_delete=models.CASCADE)
     class_type = models.IntegerField(choices=ClassType.list())
     mark = models.CharField(max_length=30, null=True, blank=True)
+    teachers = models.ManyToManyField(Teacher, blank=True)
 
     def __str__(self):
         return f"{self.occurrence} - {ClassType(self.class_type).readable_name} ({self.mark})"
 
+    class Meta:
+        unique_together = ["occurrence", "class_type", "mark"]
```
Comment (asaf-kali) [inline academic_helper/models/course_occurrence.py:101]: Nice.

File: academic_helper/migrations/0011_classgroup_teachers.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+    dependencies = [
+        ('academic_helper', '0010_auto_20200629_1808'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='classgroup',
+            name='teachers',
+            field=models.ManyToManyField(blank=True, null=True, to='academic_helper.Teacher'),
```
Comment (asaf-kali) [inline academic_helper/migrations/0011_classgroup_teachers.py:16]: Remove blank and null

File: academic_helper/logic/shnaton_parser.py [hunk 3]
```diff
@@ -206,11 +200,8 @@ def parse_lecturers(lecturers):
 
     for lecturer in lecturers.contents:
         # TODO: This order is not right for course 1921!
-        if lecturer.string is not None:
+        if lecturer.string is not None and lecturer.string.strip() != '':
             ret.append(lecturer.string)
-        elif lecturer.name == "br" and lecturer.previous_sibling not in ret:
-            ret.append(ret[-1])
```
Comment (asaf-kali) [inline academic_helper/logic/shnaton_parser.py:212]: Are you sure these can be removed?

File: academic_helper/logic/shnaton_parser.py [hunk 4]
```diff
@@ -324,11 +314,24 @@ def create_course_groups(
     group, created = ClassGroup.objects.get_or_create(
         occurrence=occurrence, class_type=group_class_type, mark=group_mark
     )
+    if teachers and created:
+        group.teachers.add(*teachers)
+    if not created and teachers and set(teachers) != set(group.teachers.all()):
+        group.teachers.set(teachers)
+        log.info(f"Group {group.id} was updated")
```
Comment (asaf-kali) [inline academic_helper/logic/shnaton_parser.py:321]: I think after these you need to call `group.save()`, right?

File: academic_helper/models/course_occurrence.py (hunk 1 above)
Comment (asaf-kali) [inline academic_helper/models/course_occurrence.py:101]: Looks like you forgot the migration for that. You can just delete migration `0011` and run again `makemigrations`. Your database will then wont work, just delete it and run `migrate`.

File: academic_helper/logic/shnaton_parser.py (hunk 4 above)
Comment (matanel-oren) [reply academic_helper/logic/shnaton_parser.py:321]: I don't think, it was created before.

**Comment by matanel-oren** [comment]
all of your comments were resolved

**Comment by asaf-kali** [comment]
@matanel-oren Something here has changed, and now the test `test_simple_base_models` fails. It receives `13` classes instead of `14` classes. Can you check it out?

**Comment by matanel-oren** [comment]
Yes, it's not a bug, it's a feature.. It's related to the "מועדים מיוחדים" issue . Until we refer that issue, this is the expected behavior. Is't because two classes have the same day and hour (and not the date), hence for us they are the same. See course 96203, 4th and 10th classes. Actually if this test was pass it was a bug.
I hope I will have time next week to try adding the handling of "מועדים מיוחדים" cases.

**Comment by asaf-kali** [comment]
I just added מועדים מיוחדים, and it passes. Nice catch.

**Comment by matanel-oren** [comment]
Because it assigned to them a different teachers, and this PR do exactly this - doesn't let you add the same class with different teacher as different classes.

**Pull request merged**